### PR TITLE
Objective wrapper

### PR DESCRIPTION
**Pull request by HenrikMettler**
Objective wrapper that checks `if individual.fitness is None` and returns `objective(individual, *args)` or the `individual` otherwise.

### CODE REVIEW

File: cgp/utils.py [hunk 1]
```diff
@@ -201,3 +201,25 @@ def primitives_from_class_names(primitives_str: Tuple[str, ...]) -> Tuple[Type[N
         primitives.append(primitives_dict[s])
 
     return tuple(primitives)
+
+
+def objective(
+    f: Callable[[IndividualBase, Any, Any], IndividualBase]
+) -> Callable[[IndividualBase, Any, Any], IndividualBase]:
+    """ Decorator for the objective function.
+    Skips evaluations of individuals with fitness != None
```
Comment (mschmidt87) [inline cgp/utils.py:210]: perhaps add one more half-sentence like I suggested.

Comment (mschmidt87) [inline cgp/utils.py:210]:
```suggestion
    Skips evaluations of individuals with fitness != None, i.e., for individuals where the fitness has been evaluated already.
```

File: cgp/utils.py [hunk 2]
```diff
@@ -201,3 +201,25 @@ def primitives_from_class_names(primitives_str: Tuple[str, ...]) -> Tuple[Type[N
         primitives.append(primitives_dict[s])
 
     return tuple(primitives)
+
+
```
Comment (mschmidt87) [inline cgp/utils.py:205]: The `mypy` issue can be solved with these suggestions:

File: cgp/utils.py [hunk 3]
```diff
@@ -201,3 +201,25 @@ def primitives_from_class_names(primitives_str: Tuple[str, ...]) -> Tuple[Type[N
...
+def objective(
+    f: Callable[[IndividualBase, Any, Any], IndividualBase]
+) -> Callable[[IndividualBase, Any, Any], IndividualBase]:
```
Comment (mschmidt87) [inline cgp/utils.py:208]:
```suggestion
F = TypeVar('F', bound=Callable[..., IndividualBase])

def objective(f: F) -> F:
```

File: cgp/utils.py (hunk 2 above)
Comment (mschmidt87) [reply cgp/utils.py:205]: you'll need to import `TypeVar, cast` from `typing`

File: cgp/utils.py [hunk 4]
```diff
@@ -201,3 +201,25 @@ def primitives_from_class_names(primitives_str: Tuple[str, ...]) -> Tuple[Type[N
...
+    """
+
+    @functools.wraps(f)
+    def wrapped(individual: IndividualBase, *args: Any, **kwargs) -> IndividualBase:
+        if individual.fitness is None:
+            return f(individual, *args, **kwargs)
+        else:
+            return individual
+
+    return wrapped
```
Comment (mschmidt87) [inline cgp/utils.py:225]:
```suggestion
    return cast(F, wrapped)
```

**Review by mschmidt87 — CHANGES_REQUESTED** [review]
This is very neat. 

I made one minor suggestion for the docstring and a suggestion to solve the mypy issue.

**Review by mschmidt87 — APPROVED** [review]
LGTM

**Comment by jakobj** [comment]
@mschmidt87 what about the discussion in #244?

**Comment by mschmidt87** [comment]
See the discussion in #244.

**Pull request closed without merging**